### PR TITLE
Add business validation unit tests

### PR DESCRIPTION
**Pull request by BjarkeMeier**
## Description
Part 1 of business validation unit tests.

## References

* #0000

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeCommandReceiver/Startup.cs [hunk 1]
```diff
@@ -64,6 +66,20 @@ public override void Configure([NotNull] IFunctionsHostBuilder builder)
...
             builder.Services.AddScoped<IChargeCommandAcceptedEventFactory, ChargeCommandAcceptedEventFactory>();
             builder.Services.AddScoped<IChargeCommandRejectedEventFactory, ChargeCommandRejectedEventFactory>();
+
+            AddIso8601Services(builder.Services);
+        }
+
+        private static void AddIso8601Services(IServiceCollection services)
+        {
+            const string timeZoneIdString = "LOCAL_TIMEZONENAME";
+            var timeZoneId = Environment.GetEnvironmentVariable(timeZoneIdString) ??
```
Comment (bemwen) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeCommandReceiver/Startup.cs:76]: We need to extend the charge command receiver to include this setting in terraform

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/BusinessValidation/Rules/StartDateVr209ValidationRule.cs [hunk 2]
```diff
@@ -13,25 +13,41 @@
...
+            [NotNull] StartDateVr209ValidationRuleConfiguration configuration,
+            [NotNull] IZonedDateTimeService zonedDateTimeService)
         {
             _validityStartDate = command.MktActivityRecord!.ValidityStartDate;
 
-            _startOfValidInterval = command.RequestDate.Plus(Duration.FromDays(configuration.ValidIntervalFromNowInDays.Start));
-            _endOfValidInterval = command.RequestDate.Plus(Duration.FromDays(configuration.ValidIntervalFromNowInDays.End));
+            var today = zonedDateTimeService.GetZonedDateTime(command.RequestDate).Date;
+            _periodStart = CalculatePeriodPoint(configuration.ValidIntervalFromNowInDays.Start, zonedDateTimeService, today);
+            _periodEnd = CalculatePeriodPoint(configuration.ValidIntervalFromNowInDays.End + 1, zonedDateTimeService, today);
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/BusinessValidation/Rules/StartDateVr209ValidationRule.cs:37]: Why does the PeriodEnd have +1 added to it?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/BusinessValidation/BusinessUpdateValidationRulesFactoryTests.cs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    {
+        [Theory]
+        [InlineAutoDomainData]
+        public async Task CreateRulesForUpdateCommandAsync_ReturnsRulesForTariffUpdateCommand(
+            [NotNull][Frozen] Mock<IUpdateRulesConfigurationRepository> updateRulesConfigurationRepository,
+            [NotNull] BusinessUpdateValidationRulesFactory sut,
+            [NotNull] ChargeCommand chargeCommand)
+        {
+            // Arrange
+            var configuration = CreateConfiguration();
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/BusinessValidation/BusinessUpdateValidationRulesFactoryTests.cs:43]: Perhaps move the creation of the configuration to a private method.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/ChargeCommandValidationResultTests.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using GreenEnergyHub.Charges.Application.Validation;
+using GreenEnergyHub.Charges.Application.Validation.BusinessValidation;
+using Xunit;
+
+namespace GreenEnergyHub.Charges.Tests.Application.Validation
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/ChargeCommandValidationResultTests.cs:21]: for consistency, even if tests are so small we should have arrange / act / assert in my opinion.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/ChargeCommandValidatorTests.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        public async Task ValidateAsync_WhenInputValidationSucceedsAndBusinessValidationFails_ReturnsInvalid(
+            [NotNull] [Frozen] Mock<IChangeOfChargeTransactionInputValidator> inputValidator,
+            [NotNull][Frozen] Mock<IChargeCommandBusinessValidator> businessValidator,
+            [NotNull] ChargeCommandValidator sut,
+            ChargeCommand anyCommand)
+        {
+            // Arrange
+            var validResult = ChargeCommandValidationResult.CreateSuccess();
+            var invalidResult = CreateInvalidValidationResult();
+            inputValidator
```
Comment (lasrinnil) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/ChargeCommandValidatorTests.cs:62]: move the setup of the business- and inputvalidation to private methods.

**Review by lasrinnil — APPROVED** [review]

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeCommandReceiver/Startup.cs (hunk 1 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.ChargeCommandReceiver/Startup.cs:76]: Good catch. Resolved.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/BusinessValidation/Rules/StartDateVr209ValidationRule.cs (hunk 2 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Application/Validation/BusinessValidation/Rules/StartDateVr209ValidationRule.cs:37]: It is to ensure that all of the last day (until midnight) is included in the interval. This is also verified by the unit tests.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/BusinessValidation/BusinessUpdateValidationRulesFactoryTests.cs (hunk 3 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/BusinessValidation/BusinessUpdateValidationRulesFactoryTests.cs:43]: Isn't it in a private method?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/ChargeCommandValidationResultTests.cs (hunk 4 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/ChargeCommandValidationResultTests.cs:21]: Personally I'm in favour of the old decision made in the ART. You can read it [here](https://dev.azure.com/energinet/Datahub/_wiki/wikis/DataHub/3711/Projects-and-Coding?anchor=unit-test-structure-***tbd***). For reference it says:

- As long as there are less than three lines of code in a test, they appear without any special formatting or white space.
- When a test contains more than three lines of code, you separate the three phases with a blank line.
- When a single phase contains so many lines of code that you'll need to divide it into subsections to make it readable, you should explicitly mark the beginning of each phase with a code comment.

Perhaps we should consider putting this in our code-of-conduct.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/BusinessValidation/BusinessUpdateValidationRulesFactoryTests.cs (hunk 3 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Application/Validation/BusinessValidation/BusinessUpdateValidationRulesFactoryTests.cs]: We have made a refactoring.

File: build/infrastructure/azfun-charges-chargecommandreceiver.tf [hunk 6]
```diff
@@ -35,6 +35,7 @@ module "azfun_charges_charge_command_receiver" {
     COMMAND_REJECTED_TOPIC_NAME                  = module.sbt_command_rejected.name
     COMMAND_RECEIVED_SUBSCRIPTION_NAME           = azurerm_servicebus_subscription.sbs_command_received.name
     CHARGE_DB_CONNECTION_STRING                  = local.CHARGE_DB_CONNECTION_STRING
+    LOCAL_TIMEZONENAME                           = "Europe/Copenhagen"
```
Comment (bemwen) [inline build/infrastructure/azfun-charges-chargecommandreceiver.tf:38]: Consider moving this to local.tf like we had in the archived repo.

To have "magic strings" as few places as possible in terraform and not hidden away in individual files

**Review by bemwen — APPROVED** [review]

**Pull request merged**